### PR TITLE
Reduce unnecessary asserts #239

### LINKED ISSUE #239: Reduce unnecessary asserts
We need to define where to use asserts and where to omit them. For example, asserting every parameter to every method as not null may be unnecessary.

### PR DESCRIPTION
**Pull request by Zhiyuan-Amos**
Fixes #239 

Proposed merge commit message:
```
Parameter validation is done by using asserts, and throws AssertionError
when validation fails.

Asserts may not always be enabled. Furthermore, specific exceptions
such as NullPointerException should be thrown to make the error more
explicit.

Let's
    - Replace the asserts for public APIs with checkNotNull() and
      checkArgument(), which throws an appropriate exception when 
      validation fails.
    - Enhance CollectionUtil#isAnyNull(…) to do the same operation as
      Objects#requireNonNull(T), and also able to take in multiple 
      parameters.
```

### CODE REVIEW

**Comment by CanIHasReview** [comment]
# v1

@Zhiyuan-Amos submitted v1 for review.

* [v1 1/2] [Remove unnecessary asserts](https://github.com/se-edu/addressbook-level4/pull/414/commits/207cf4b9f21fe006b2fc76e35400d71fa41df807)
* [v1 2/2] [Refactor asserts used for parameter validation](https://github.com/se-edu/addressbook-level4/pull/414/commits/c1d7c23902f90c66081a903272f0e737bc5caa7a)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/0f2cb22b3dbf7cdc592e2b58261728241f4d4166...c1d7c23902f90c66081a903272f0e737bc5caa7a))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/414/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: src/main/java/seedu/address/model/person/UniquePersonList.java [hunk 1]
```diff
@@ -25,7 +25,6 @@
      * Returns true if the list contains an equivalent person as the given argument.
      */
     public boolean contains(ReadOnlyPerson toCheck) {
-        assert toCheck != null;
```
Comment (yamgent) [inline src/main/java/seedu/address/model/person/UniquePersonList.java]: Does `ObservableList` allow `null` elements? If it does, then `internalLIst.contains()` will not trigger an NPE.

Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/model/person/UniquePersonList.java]: you are right, it allows null elements. If the list allows for null elements, does it mean that we should allow the method to be called `contains(null)`?

Comment (yamgent) [inline src/main/java/seedu/address/model/person/UniquePersonList.java]: For our side, it should prohibit accepting `null`, so we should use `checkNotNull()`

**Comment by CanIHasReview** [comment]
# v2

@Zhiyuan-Amos submitted v2 for review.

* [v2 1/1] [Replace asserts with proper exceptions for parameter validation](https://github.com/se-edu/addressbook-level4/pull/414/commits/96170277a1184ad6ceb3ba9cc780992d04e4fe07)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/0f2cb22b3dbf7cdc592e2b58261728241f4d4166...96170277a1184ad6ceb3ba9cc780992d04e4fe07)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/compare/3f03acefefce197fe60ae063bd5b9c6e83391338~1...3f03acefefce197fe60ae063bd5b9c6e83391338))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/414/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by yamgent** [comment]
@pyokagan @MightyCupcakes For your review please, as both of you have previously suggested a guideline on how we can replace the asserts. :)

**Comment by pyokagan** [comment]
Check if any of the comments from #300 apply here.

**Comment by pyokagan** [comment]
It may be a good idea to start off from the commits in #300, rather than starting from scratch (so reviewers don't have to make the same comments).

**Comment by Zhiyuan-Amos** [comment]
Yup I've referred to the commits from #300. We discussed with prof, prof's opinion is that we should replace `assert` with `checkNotNull` for every public method, even though there may be methods inside the method that will throw NPE. 

Rationale is: Some of the method calls that throw NPE are found in the middle of the execution of the method. The code executed earlier in the method may have persistent effects, which may result in unintended changes to the program when the app crashes due to NPE.

**Comment by pyokagan** [comment]
@Zhiyuan-Amos 
> we should replace assert with checkNotNull for every public method, even though there may be methods inside the method that will throw NPE.

OK. That should have been in the commit message.

How about other differences between this PR and #300? E.g., the use of Guava's `checkNotNull()` instead of JDK's `requireNotNull()`? As well as the other stylistic differences like the use of the:
```java
this.someField = requireNonNull(someField);
```
pattern?

File: src/main/java/seedu/address/model/ModelManager.java [hunk 2]
```diff
@@ -30,7 +33,7 @@
      */
     public ModelManager(ReadOnlyAddressBook addressBook, UserPrefs userPrefs) {
         super();
-        assert !CollectionUtil.isAnyNull(addressBook, userPrefs);
+        checkArgument(!CollectionUtil.isAnyNull(addressBook, userPrefs));
```
Comment (pyokagan) [inline src/main/java/seedu/address/model/ModelManager.java]: Should this be throwing an `NullPointerException` or an `IllegalArgumentException`?

Comment (pyokagan) [inline src/main/java/seedu/address/model/ModelManager.java]: Probably `NullPointerException`, I think, since this is implicitly just checking the arguments to this constructor.

**Comment by pyokagan** [comment]
> the use of Guava's checkNotNull() instead of JDK's requireNotNull()?

I think that it's not good to use Guava's functions actually, as checking arguments and not-null has to be done *throughout* the code-base and it would mean that the code in addressbook-level4 can't be trivially shared with the other levels (since they are not supposed to have access to a java build system like gradle). Instead, I think it's better to just use the standard library's `requireNotNull()` and write our own `checkArgument()`.

File: src/main/java/seedu/address/model/ModelManager.java (hunk 2 above)
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/model/ModelManager.java]: Yup this is a good catch. Perhaps `CollectionUtil.isAnyNull` should throw `NullPointerException` if any of the variables are null? 

Currently `CollectionUtil.isAnyNull` is only used in `assert` statements so when assert is enabled, they throw `AssertionError`. As such, if we let `CollectionUtil.isAnyNull` throw `NPE` instead, I don't think there's much that need to be changed in the codebase :>

**Comment by CanIHasReview** [comment]
# v3

@Zhiyuan-Amos submitted v3 for review.

* [v3 1/2] [Replace asserts with proper exceptions for parameter validation](https://github.com/se-edu/addressbook-level4/pull/414/commits/9f1ed0f6cab034abcb3a121c75d39a523f522477)
* [v3 2/2] [Update CollectionUtil\#isAnyNull to throw NullPointerException](https://github.com/se-edu/addressbook-level4/pull/414/commits/fc13a6090ff4d8acd2461d6fb12e290ad90da671)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/ef06553753b5073c012ccac9ab1082c46b9d0f51...fc13a6090ff4d8acd2461d6fb12e290ad90da671)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/compare/266f62f610762923c4e9f81a272b82b3ae96b37e~1...266f62f610762923c4e9f81a272b82b3ae96b37e))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/414/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: src/main/java/seedu/address/commons/util/AppUtil.java [hunk 3]
```diff
@@ -9,8 +11,29 @@
...
+    }
+
+    /**
+     * Checks that {@code condition} is true. Used for validating arguments to methods.
+     *
+     * @throws IllegalArgumentException with {@code errorMessage} if {@code condition} is false.
+     */
+    public static void checkArgument(Boolean condition, String errorMessage) {
+        if (!condition) {
+            throw new IllegalArgumentException();
```
Comment (limmlingg) [inline src/main/java/seedu/address/commons/util/AppUtil.java]: `errorMessage` isn't used.

File: src/main/java/seedu/address/model/ModelManager.java (hunk 2 above)
Comment (yamgent) [inline src/main/java/seedu/address/model/ModelManager.java]: ~~@Zhiyuan-Amos I think this still doesn't throw `NullPointerException`?~~

File: src/test/java/guitests/guihandles/AlertDialogHandle.java [hunk 4]
```diff
@@ -16,7 +18,7 @@ public AlertDialogHandle(GuiRobot guiRobot, Stage primaryStage, String dialogTit
     }
 
     public boolean isMatching(String headerMessage, String contentMessage) {
-        assert intermediateStage.isPresent() : "Alert dialog is not present";
+        checkArgument(intermediateStage.isPresent());
```
Comment (yamgent) [inline src/test/java/guitests/guihandles/AlertDialogHandle.java]: This one should use `checkArgument()` with the message.

File: src/test/java/seedu/address/commons/util/CollectionUtilTest.java [hunk 5]
```diff
@@ -107,6 +98,46 @@ public void elementsAreUnique() throws Exception {
         assertNotUnique(null, "a", "b", null);
     }
 
+    private void assertNullPointerExceptionThrown(Object... objects) {
+        try {
+            requireNonNull(objects);
+            fail();
+        } catch (NullPointerException npe) {
```
Comment (yamgent) [inline src/test/java/seedu/address/commons/util/CollectionUtilTest.java]: Use a comment if it is intended behavior (otherwise checkstyle will complain about empty catch block).

```
} catch (NullPointerException npe) {
    // expected behavior
}
```
http://checkstyle.sourceforge.net/config_blocks.html#EmptyCatchBlock

File: src/test/java/seedu/address/commons/util/CollectionUtilTest.java [hunk 6]
```diff
@@ -107,6 +98,46 @@ public void elementsAreUnique() throws Exception {
...
+        } catch (Exception e) {
+            fail();
```
Comment (yamgent) [inline src/test/java/seedu/address/commons/util/CollectionUtilTest.java]: No need for this, just let the exception throw.

File: src/test/java/seedu/address/commons/util/CollectionUtilTest.java [hunk 7]
```diff
@@ -107,6 +98,46 @@ public void elementsAreUnique() throws Exception {
...
+    }
+
+    private void assertNullPointerExceptionNotThrown(Object... objects) {
+        try {
+            requireNonNull(objects);
+            return;
+        } catch (NullPointerException npe) {
+            fail();
+        }
+        fail();
```
Comment (yamgent) [inline src/test/java/seedu/address/commons/util/CollectionUtilTest.java]: 1. Do we need a try-catch block for this? Seems like you can just let the exception throw?
2. Even if we need try-catch, we don't need the `return;` on line 124 and `fail();` on line 128:

```
try {
    requireNonNull(objects);
} catch (NullPointerException npe) {
    fail();
}
```

File: src/test/java/seedu/address/commons/util/CollectionUtilTest.java [hunk 8]
```diff
@@ -2,84 +2,75 @@
...
-        assertFalse(CollectionUtil.isAnyNull());
+        assertNullPointerExceptionNotThrown();
 
         // any non-empty argument list
-        assertFalse(CollectionUtil.isAnyNull(new Object(), new Object()));
-        assertFalse(CollectionUtil.isAnyNull("test"));
-        assertFalse(CollectionUtil.isAnyNull(""));
+        assertNullPointerExceptionNotThrown(new Object(), new Object());
+        assertNullPointerExceptionNotThrown("test");
+        assertNullPointerExceptionNotThrown("");
```
Comment (yamgent) [inline src/test/java/seedu/address/commons/util/CollectionUtilTest.java]: What's the difference between `"test"` and `""`?

File: src/test/java/seedu/address/commons/util/CollectionUtilTest.java [hunk 9]
```diff
@@ -2,84 +2,75 @@
...
 
         // argument lists with just one null at the beginning
-        assertTrue(CollectionUtil.isAnyNull((Object) null));
-        assertTrue(CollectionUtil.isAnyNull(null, "", new Object()));
-        assertTrue(CollectionUtil.isAnyNull(null, new Object(), new Object()));
+        assertNullPointerExceptionThrown((Object) null);
```
Comment (yamgent) [inline src/test/java/seedu/address/commons/util/CollectionUtilTest.java]: should be under `// null reference`?

File: src/test/java/seedu/address/commons/util/CollectionUtilTest.java [hunk 10]
```diff
@@ -2,84 +2,75 @@
...
-        assertTrue(CollectionUtil.isAnyNull("", new Object(), null));
-        assertTrue(CollectionUtil.isAnyNull(new Object(), new Object(), null));
+        assertNullPointerExceptionThrown("", new Object(), null);
+        assertNullPointerExceptionThrown(new Object(), new Object(), null);
+
+        // null reference
+        assertNullPointerExceptionThrown((Object[]) null);
 
         // confirms nulls inside lists in the argument list are not considered
         List<Object> containingNull = Arrays.asList((Object) null);
```
Comment (yamgent) [inline src/test/java/seedu/address/commons/util/CollectionUtilTest.java]: `listContainingNull`?

File: src/test/java/seedu/address/commons/util/CollectionUtilTest.java (hunk 8 above)
Comment (Zhiyuan-Amos) [inline src/test/java/seedu/address/commons/util/CollectionUtilTest.java]: erm i have no idea. can we don't clean up the test in this PR? :P

File: src/test/java/seedu/address/commons/util/CollectionUtilTest.java (hunk 9 above)
Comment (Zhiyuan-Amos) [inline src/test/java/seedu/address/commons/util/CollectionUtilTest.java]: same comment as above :P

File: src/test/java/seedu/address/commons/util/CollectionUtilTest.java (hunk 10 above)
Comment (Zhiyuan-Amos) [inline src/test/java/seedu/address/commons/util/CollectionUtilTest.java]: same comment as above :P

File: src/test/java/seedu/address/commons/util/CollectionUtilTest.java (hunk 8 above)
Comment (yamgent) [inline src/test/java/seedu/address/commons/util/CollectionUtilTest.java]: Ok, can be a separate PR

**Comment by CanIHasReview** [comment]
# v4

@Zhiyuan-Amos submitted v4 for review.

* [v4 1/2] [Replace asserts with proper exceptions for parameter validation](https://github.com/se-edu/addressbook-level4/pull/414/commits/db865acf7dd3996230d0a2cf35672d450b5b3cc6)
* [v4 2/2] [Update CollectionUtil\#isAnyNull\(Object\.\.\.\) to throw NPE](https://github.com/se-edu/addressbook-level4/pull/414/commits/16b29b35dd630821077313cdfc43e5ba2bbec43d)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/cc22ed642b34f15798facec1bddd597aa8c0c395...16b29b35dd630821077313cdfc43e5ba2bbec43d)) ([:chart_with_upwards_trend: Interdiff between v3 and v4](https://github.com/se-edu/addressbook-level4/compare/34f907792fc92edcd4abed29024ca5f5ae685b70~1...34f907792fc92edcd4abed29024ca5f5ae685b70))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/414/4/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: src/test/java/seedu/address/commons/util/CollectionUtilTest.java [hunk 11]
```diff
@@ -107,6 +98,32 @@ public void elementsAreUnique() throws Exception {
...
+            fail();
+        } catch (NullPointerException npe) {
+            // expected behavior
+        }
+    }
+
+    private void assertNullPointerExceptionThrown(Collection<?> objects) {
+        try {
+            requireNonNull(objects);
+            fail();
```
Comment (yamgent) [inline src/test/java/seedu/address/commons/util/CollectionUtilTest.java]: Use `fail(String)`.

**Comment by CanIHasReview** [comment]
# v5

@Zhiyuan-Amos submitted v5 for review.

* [v5 1/2] [Replace asserts with proper exceptions for parameter validation](https://github.com/se-edu/addressbook-level4/pull/414/commits/4271d5b094b251cb67fe869f59edc496d8efe0c6)
* [v5 2/2] [Update CollectionUtil\#isAnyNull\(Object\.\.\.\) to throw NPE](https://github.com/se-edu/addressbook-level4/pull/414/commits/7aacef5e3d82d1b373d08c6e3ccf63267816b0fd)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/6af393367c998910784881146ffa049725bc31bd...7aacef5e3d82d1b373d08c6e3ccf63267816b0fd)) ([:chart_with_upwards_trend: Interdiff between v4 and v5](https://github.com/se-edu/addressbook-level4/compare/e46bcc70f2c25bd0121113f1179eec3be4746948~1...e46bcc70f2c25bd0121113f1179eec3be4746948))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/414/5/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by Zhiyuan-Amos** [comment]
@damithc done

File: src/main/java/seedu/address/ui/UiPart.java [hunk 12]
```diff
@@ -27,8 +29,7 @@
      * The FXML file must not specify the {@code fx:controller} attribute.
      */
     public UiPart(URL fxmlFileUrl) {
-        assert fxmlFileUrl != null;
-        fxmlLoader = new FXMLLoader(fxmlFileUrl);
+        fxmlLoader = new FXMLLoader(requireNonNull(fxmlFileUrl));
```
Comment (MightyCupcakes) [inline src/main/java/seedu/address/ui/UiPart.java]: Some inconsistencies - is there a reason why you are in-lining `requireNonNull` here while the others separate the check into another line like so:

```java
requireNonNull(toRemove);
final boolean personFoundAndDeleted = internalList.remove(toRemove);
```

Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/ui/UiPart.java]: Yup, we only in-line when we are assigning to **instance** variable on the first line. :)

**Comment by CanIHasReview** [comment]
# v6

@Zhiyuan-Amos submitted v6 for review.

* [v6 1/2] [Replace asserts with proper exceptions for parameter validation](https://github.com/se-edu/addressbook-level4/pull/414/commits/6ed98d61030e184b7229a740693639c251d89f2b)
* [v6 2/2] [Update CollectionUtil\#isAnyNull\(Object\.\.\.\) to throw NPE](https://github.com/se-edu/addressbook-level4/pull/414/commits/422887597ce23aceb9907bc2c01509ed2684b0fa)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/6af393367c998910784881146ffa049725bc31bd...422887597ce23aceb9907bc2c01509ed2684b0fa)) ([:chart_with_upwards_trend: Interdiff between v5 and v6](https://github.com/se-edu/addressbook-level4/compare/0eab73b516844aa04aee62148c3f76ef9342ca3a~1...0eab73b516844aa04aee62148c3f76ef9342ca3a))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/414/6/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 13]
```diff
@@ -53,11 +55,9 @@
      * @param editPersonDescriptor details to edit the person with
      */
     public EditCommand(int filteredPersonListIndex, EditPersonDescriptor editPersonDescriptor) {
-        assert filteredPersonListIndex > 0;
-        assert editPersonDescriptor != null;
-
+        checkArgument(filteredPersonListIndex > 0);
+        this.editPersonDescriptor = new EditPersonDescriptor(requireNonNull(editPersonDescriptor));
```
Comment (MightyCupcakes) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: How about this? This is not a direct assignment too.

Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: @yamgent how ah? This one looks quite lame leh to not inline it. But based on current standards we should inline. 

Thanks for the catch Jeremy :>

Comment (yamgent) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: Don't inline this then. Let's stick to the convention that only direct assignment = inline, that way it is easier to enforce the convention.

Comment (yamgent) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: I think the standards can be as follow:
* If it consist of all direct assignment, it is allowed.
* If there is only one parameter, and is passed into the constructor in the first non-precondition line, and the constructor only takes in one parameter, it is allowed.

Comment (yamgent) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: cc. @damithc would that be fine, or would the standard become too hard to enforce?

Comment (MightyCupcakes) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: Point number 2 of sounds really convoluted. Standards should be concise so that incoming developers can easily understand.

Comment (damithc) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: I prefer the simpler one too. It's easier to enforce/review when the parameter checking happens at the beginning of the method. A shortcut can be allowed only for very clear cut cases where the whole method is just assignment statements.

**Comment by CanIHasReview** [comment]
# v7

@Zhiyuan-Amos submitted v7 for review.

* [v7 1/2] [Replace asserts with proper exceptions for parameter validation](https://github.com/se-edu/addressbook-level4/pull/414/commits/28c175b7cf7b877c81dbe05c1dcdc41561b9543e)
* [v7 2/2] [Update CollectionUtil\#isAnyNull\(Object\.\.\.\) to throw NPE](https://github.com/se-edu/addressbook-level4/pull/414/commits/a4fa2dbcfc7407927c6ba1292a5ec421b4ae3c81)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/cfc8ecc72a73a989ecc58427e6ddd1c564361408...a4fa2dbcfc7407927c6ba1292a5ec421b4ae3c81)) ([:chart_with_upwards_trend: Interdiff between v6 and v7](https://github.com/se-edu/addressbook-level4/compare/826c5f79287e7549a1f8b55db5be019fbe3eb2c3~1...826c5f79287e7549a1f8b55db5be019fbe3eb2c3))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/414/7/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: src/main/java/seedu/address/logic/commands/EditCommand.java (hunk 13 above)
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: alright I have done the changes!

**Comment by CanIHasReview** [comment]
# v8

@Zhiyuan-Amos submitted v8 for review.

* [v8 1/2] [Replace asserts with proper exceptions for parameter validation](https://github.com/se-edu/addressbook-level4/pull/414/commits/b71e90e76d0c56590088b3d4a84d92d7b54d2f59)
* [v8 2/2] [Update CollectionUtil\#isAnyNull\(Object\.\.\.\) to throw NPE](https://github.com/se-edu/addressbook-level4/pull/414/commits/fc4b49f3f473115bdc981a12baaa1a5f675fb23c)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/cfc8ecc72a73a989ecc58427e6ddd1c564361408...fc4b49f3f473115bdc981a12baaa1a5f675fb23c)) ([:chart_with_upwards_trend: Interdiff between v7 and v8](https://github.com/se-edu/addressbook-level4/compare/48cafed10b96781fd29256d7432b2d2bcc9bdb06~1...48cafed10b96781fd29256d7432b2d2bcc9bdb06))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/414/8/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by CanIHasReview** [comment]
# v9

@Zhiyuan-Amos submitted v9 for review.

* [v9 1/2] [Replace asserts with proper exceptions for parameter validation](https://github.com/se-edu/addressbook-level4/pull/414/commits/a37629cf4c31f6783ffefbb4cd85bbcf3bc0f016)
* [v9 2/2] [Update CollectionUtil\#isAnyNull\(Object\.\.\.\) to throw NPE](https://github.com/se-edu/addressbook-level4/pull/414/commits/202c82428d993a976d13eb1a5633119ef6c951d2)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/cfc8ecc72a73a989ecc58427e6ddd1c564361408...202c82428d993a976d13eb1a5633119ef6c951d2)) ([:chart_with_upwards_trend: Interdiff between v8 and v9](https://github.com/se-edu/addressbook-level4/compare/8b8fb6e641f0c1de40de18dc143f74bc0be7e84a~1...8b8fb6e641f0c1de40de18dc143f74bc0be7e84a))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/414/9/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by Zhiyuan-Amos** [comment]
added tests for AppUtil#checkArgument

File: src/main/java/seedu/address/commons/util/CollectionUtil.java [hunk 14]
```diff
@@ -12,17 +12,18 @@
...
     }
 
     /**
-     * Returns true if any element of {@code items} is null.
-     * @throws NullPointerException if {@code items} itself is null.
+     * Throws NullPointerException if {@code items} or any element of {@code items} is null.
      */
-    public static boolean isAnyNull(Collection<?> items) {
-        return items.stream().anyMatch(Objects::isNull);
+    public static void requireNonNull(Collection<?> items) {
```
Comment (pyokagan) [inline src/main/java/seedu/address/commons/util/CollectionUtil.java]: Giving this the same name as `Objects#requireNonNull` may be dangerous though. If `import static seedu.address.commons.util.CollectionUtil.requireNonNull` is not present in the file, then `Objects#requireNonNull` would be used instead, which would be hard to detect by just reading the diff in a code review.

Maybe give this another name? Maybe `requireNonNullElements` or `requireAllNonNull`.

Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/commons/util/CollectionUtil.java]: alright!

**Comment by Zhiyuan-Amos** [comment]
@MightyCupcakes any comments? :>

**Comment by MightyCupcakes** [comment]
No other comments.

**Comment by CanIHasReview** [comment]
# v10

@Zhiyuan-Amos submitted v10 for review.

* [v10 1/2] [Replace asserts with proper exceptions for parameter validation](https://github.com/se-edu/addressbook-level4/pull/414/commits/a37629cf4c31f6783ffefbb4cd85bbcf3bc0f016)
* [v10 2/2] [Update CollectionUtil\#isAnyNull\(Object\.\.\.\) to throw NPE](https://github.com/se-edu/addressbook-level4/pull/414/commits/288a958a8d077613b9eed81e5f20219cc4d97449)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/cfc8ecc72a73a989ecc58427e6ddd1c564361408...288a958a8d077613b9eed81e5f20219cc4d97449)) ([:chart_with_upwards_trend: Interdiff between v9 and v10](https://github.com/se-edu/addressbook-level4/compare/8cb2f48d590ac66f1574239f4714c5eb953e886a~1...8cb2f48d590ac66f1574239f4714c5eb953e886a))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/414/10/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by Zhiyuan-Amos** [comment]
@MightyCupcakes ping :>
